### PR TITLE
MODLD-339: Derive marc 010 from LCCN Resource description

### PR DESCRIPTION
**Pull request by askhat-abishev**

### CODE REVIEW

**Review by aleksei-pronichev — APPROVED** [review]

File: src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+          subfield.set(A);
+        } else if (link.equals(CANCINV)) {
+          subfield.set(Z);
+        }
+      }, () -> subfield.set(A));
+    return subfield.get() != null ? Optional.of(subfield.get()) : Optional.empty();
+  }
+
+  private String getPropertyValue(PropertyDictionary property, Resource resource) {
+    return resource.getDoc().get(property.getValue()).get(0).asText();
```
Comment (aleksei-pronichev) [inline src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java:70]: minor. to many chances for nullpointer
```
Optional.of (resource)
.map(Resource::getDoc)
...
```

File: src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  @Override
+  public List<DataField> map(Resource resource) {
+    var dataField = marcFactory.newDataField(TAG, EMPTY, EMPTY);
+    getSubfield(resource)
+      .ifPresent(s -> dataField.addSubfield(marcFactory.newSubfield(s, getPropertyValue(NAME, resource))));
+    return List.of(dataField);
+  }
+
+  private Optional<Character> getSubfield(Resource resource) {
+    final var subfield = new AtomicReference<Character>();
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java:53]: I think we can avoid the use of `AtomicReference` if we do as below-

```
  private Optional<Character> getSubfield(Resource resource) {
    return resource.getOutgoingEdges().stream()
      .filter(resourceEdge -> resourceEdge.getPredicate() == STATUS)
      .findFirst()
      .map(e -> getPropertyValue(LINK, e.getTarget()))
      .map(link -> switch (link) {
        case CURRENT -> Optional.of(A);
        case CANCINV -> Optional.of(Z);
        default -> Optional.<Character>empty();
      })
      .orElse(Optional.of(A));
  }

```

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+import org.apache.commons.lang3.StringUtils;
+import org.folio.marc4ld.service.ld2marc.processing.combine.DataFieldCombiner;
+import org.marc4j.marc.DataField;
+import org.marc4j.marc.Subfield;
+
+public abstract class AbstractDataFieldCombiner implements DataFieldCombiner {
+
+  private static final char EMPTY = ' ';
+
+  private DataField combinedField;
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java:22]: Something to consider -

Both 245 and 010 are non-repeatable marc fields. Hence, there will be only one occurrence of `combinedField` for those 2 marc fields. Hence, a single occurrence of `combinedField` (as given in this class) is perfectly fine.

However, there are a lot of marc fields that can repeat (say, marc 020, 264 etc). In such cases, we will have to maintain a list of combinedFields (`List< DataField> combinedFields;`). 

It is difficult to design this class for such future requirements without knowing the actual requirements.  May be, we will rename this class to `AbstractNonRepeatableDataFieldCombiner` and a create different class for `RepeatableDataField` in future.

So, I am fine to leave this class like this for now. However, this is something we have to keep in mind.

File: src/test/java/org/folio/marc4ld/mapper/test/MonographTestUtil.java [hunk 4]
```diff
@@ -628,6 +620,26 @@ public static Resource createSampleWork() {
     ).setLabel("Work: label");
   }
 
+  public static Resource status(String value) {
```
Comment (pkjacob) [inline src/test/java/org/folio/marc4ld/mapper/test/MonographTestUtil.java:623]: minor - can we make this method to accept a boolean value `isCancelled`? This way, we can avoid passing string "canceled or invalid" and matching against it in line 628.

**Review by pkjacob — APPROVED** [review]

File: src/test/java/org/folio/marc4ld/mapper/test/MonographTestUtil.java (hunk 4 above)
Comment (askhat-abishev) [reply src/test/java/org/folio/marc4ld/mapper/test/MonographTestUtil.java:623]: I suppose that we pass it to use it for the label. So I've left it as is for now.

File: src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java (hunk 2 above)
Comment (askhat-abishev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/mapper/impl/LccnMapper.java]: done

File: src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+import java.util.Objects;
+import java.util.Optional;
+import org.apache.commons.lang3.StringUtils;
+import org.folio.marc4ld.service.ld2marc.processing.combine.DataFieldCombiner;
+import org.marc4j.marc.DataField;
+import org.marc4j.marc.Subfield;
+
+public abstract class AbstractDataFieldCombiner implements DataFieldCombiner {
+
+  private static final char EMPTY = ' ';
```
[the remark this answers is not in the document]
Comment (askhat-abishev) [reply src/main/java/org/folio/marc4ld/service/ld2marc/processing/combine/impl/AbstractDataFieldCombiner.java]: done

File: src/main/java/org/folio/marc4ld/util/Constants.java [hunk 6]
```diff
@@ -8,6 +8,7 @@ public class Constants {
   public static final String FIELD_UUID = "999";
   public static final char SUBFIELD_INVENTORY_ID = 'i';
   public static final char SUBFIELD_SRS_ID = 's';
+  public static final char EMPTY = ' ';
```
Comment (pkjacob) [inline src/main/java/org/folio/marc4ld/util/Constants.java:11]: I think 'SPACE' was a better name for this constant. Let us fix this in another PR.

**Pull request merged**